### PR TITLE
[BEAM-1347] Convert an InputStream into an Iterable<T> using the Beam Fn data specification

### PR DESCRIPTION
**Pull request by lukecwik**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This is towards sharing common code that supports the Beam Fn State API and the Beam Fn Data API.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @vikkyrk

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
retest this please

Last failure run was unrelated: https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/13854/

**Comment by lukecwik** [comment]
retest this please

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/stream/DataStreams.java [hunk 1]
```diff
@@ -100,6 +105,70 @@ public int read(byte[] b, int off, int len) throws IOException {
...
+    private final CountingInputStream countingInputStream;
+    private final PushbackInputStream pushbackInputStream;
+    private final Coder<T> coder;
+    private State currentState;
+    private T next;
+    public DataStreamDecoder(Coder<T> coder, InputStream inputStream) {
+      this.currentState = State.READ_REQUIRED;
+      this.coder = coder;
+      this.pushbackInputStream = new PushbackInputStream(inputStream, 1);
+      this.countingInputStream = new CountingInputStream(pushbackInputStream);
```
Comment (vikkyrk) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/stream/DataStreams.java]: +1 for the usage of these streams.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/stream/DataStreams.java [hunk 2]
```diff
@@ -100,6 +105,70 @@ public int read(byte[] b, int off, int len) throws IOException {
   }
 
   /**
+   * An adapter which converts an {@link InputStream} to an {@link Iterator} of {@code T} values
+   * using the specified {@link Coder}. Note that this adapter follows the Beam Fn API specification
```
Comment (vikkyrk) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/stream/DataStreams.java]: Move this `Note` to the paragraph below

**Pull request closed without merging**